### PR TITLE
Changes to handing out mutations

### PR DESCRIPTION
**Pull request by chloecat0**
## About The Pull Request
Mutations can be activated through the DNA scanner if it's been discovered or mutated from storage, giving a reason to enter the scanner. Adds an access requirement to the console so that only geneticists can distribute mutations and removes the name of the mutation from injectors so they can't just be used with no risk (a hand labeler can still be used but could be intentionally mislabeled). 
Adds wires to the DNA scanner which can disable the access requirement, irradiate, shock and toggle/lock the door.
Makes some parts of the scanner overlays so they can be locked or screwdrivered at all times.

## Why It's Good For The Game
Geneticists need to bring people into their lab or label their injectors manually which opens up using genetics as a traitor and makes taking a mutation less trustworthy. Makes current 'build console in hallway' meta less effective.

## Changelog
:cl:
add: Mutations can be activated/added through the DNA scanner
add: Added wires to the DNA scanner
tweak: Using the DNA scanner requires genetics access (unless hacked or emagged)
tweak: DNA injectors don't show their mutations
code: Discovered mutations are now an associative list
/:cl:

### CODE REVIEW

**Comment by Archanial** [comment]
Actually no, I misread some things. I like every idea besides activating mutations with scanners.

**Comment by FriendlyContractor** [comment]
I like all of the changes except ID locking the console. It would make it impossible for somebody who breaks into genetics( in the rare case geneticist refuses to make you fireball throwing hulk anyway) to steal mutations and would completly prevent players from building their own genetic labs. I belive as few things as possible on the station should be hard locked behind ID and should be possible to hack (like APCs and airlocks)

**Comment by Dingo-Dongler** [comment]
The access restriction is a bit harsh considering most other consoles (e.g. research console) don't require it. Being able to quick activate is a good idea though. Maybe you can make a much lower cooldown for injecting genes that don't exist in the host to incentivize people to enter the lab for mass mutation (as opposed to mass printing injectors).

**Review by Archanial — CHANGES_REQUESTED** [review]
Don't force antags to buy emag. Add some way to hack a console with normal tools that takes time.

**Comment by francinum** [comment]
The problem with allowing a non-antag way to disable the access lock is it will allow them to simply continue the hallway meta.

**Comment by chloecat0** [comment]
OK, I added wires to the DNA scanner which can disable the access check, so at least both the console and scanner will need to be built to give out mutations without access.

**Comment by ike709** [comment]
Can they still use the console without a scanner?

**Comment by chloecat0** [comment]
> Can they still use the console without a scanner?

Yeah it just needs the scanner for it to be hackable

**Comment by PowerfulBacon** [comment]
Can you just use the scanner without an ID if there is no linked console?

**Comment by chloecat0** [comment]
> Can you just use the scanner without an ID if there is no linked console?

The console is needed to do anything

File: code/datums/wires/dna_scanner.dm [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+	var/obj/machinery/dna_scannernew/S = holder
+	switch(wire)
+		if(WIRE_IDSCAN)
+			S.ignore_id = !S.ignore_id
+			addtimer(CALLBACK(S, /obj/machinery/dna_scannernew.proc/reset, wire), 1200)
+		if(WIRE_BOLTS)
+			S.locked = !S.locked
+			S.update_icon()
+		if(WIRE_LIMIT)
+			if(iscarbon(usr))
```
Comment (Archanial) [inline code/datums/wires/dna_scanner.dm]: you should pass mob as argument, it's better to not check usr

File: code/game/machinery/computer/dna_console.dm [hunk 2]
```diff
@@ -157,6 +158,13 @@
 	else
 		return ..()
 
+/obj/machinery/computer/scan_consolenew/interact(mob/user, special_state)
+	if(!connected_scanner?.ignore_id)
+		if(!allowed(user))
```
Comment (Archanial) [inline code/game/machinery/computer/dna_console.dm]:
```suggestion
	if(!connected_scanner?.ignore_id || !allowed(user))
```

File: code/game/machinery/computer/dna_console.dm [hunk 3]
```diff
@@ -644,6 +657,64 @@
 
 			return
 
+		//Activate mutation already in a subject's genome
+		if("activate_mutation")
+			if(!can_modify_occupant())
+				return
+
+			var/path = GET_MUTATION_TYPE_FROM_ALIAS(params["alias"])
+			if(stored_research && (path in stored_research.discovered_mutations))
```
Comment (Archanial) [inline code/game/machinery/computer/dna_console.dm]: Not sure if looking for it in list everytime is a good idea.
Make discovered mutations be associative list and just check index

File: code/game/machinery/computer/dna_console.dm [hunk 4]
```diff
@@ -1987,6 +2058,13 @@
 	if(!istype(user) || !Adjacent(user) || !user.put_in_active_hand(diskette))
 		diskette.forceMove(drop_location())
 	diskette = null
+
+/obj/machinery/computer/scan_consolenew/emag_act(mob/user)
+	obj_flags |= EMAGGED
```
Comment (Archanial) [inline code/game/machinery/computer/dna_console.dm:2061]:
```suggestion
	obj_flags &= EMAGGED
```

File: code/game/machinery/dna_scanner.dm [hunk 5]
```diff
@@ -132,13 +135,53 @@
...
 
 /obj/machinery/dna_scannernew/MouseDrop_T(mob/target, mob/user)
 	var/mob/living/L = user
-	if(user.stat || (isliving(user) && (!(L.mobility_flags & MOBILITY_STAND) || !(L.mobility_flags & MOBILITY_UI))) || !Adjacent(user) || !user.Adjacent(target) || !iscarbon(target) || !user.IsAdvancedToolUser())
+	if(user.stat || (isliving(user) && (!(L.mobility_flags & MOBILITY_STAND) || !(L.mobility_flags & MOBILITY_UI))) || !Adjacent(user) || !user.Adjacent(target) || !iscarbon(target) || !user.IsAdvancedToolUser() || locked)
 		return
 	close_machine(target)
+
+/obj/machinery/dna_scannernew/proc/irradiate(mob/living/carbon/target)
+	if(!HAS_TRAIT(target, TRAIT_RADIMMUNE))
```
Comment (Archanial) [inline code/game/machinery/dna_scanner.dm]: early return

File: code/game/machinery/dna_scanner.dm [hunk 6]
```diff
@@ -132,13 +135,53 @@
...
+		to_chat(target, "<span class='danger'>You feel warm.</span>")
+		target.rad_act(250/(damage_coeff ** 2))
+
+		if(target.has_dna() && !HAS_TRAIT(target, TRAIT_BADDNA))
+			var/resist = target.getarmor(null, "rad")
+			if(prob(max(0,100-resist)))
+				target.randmuti()
+				if(prob(20))
+					if(prob(90))
```
Comment (Archanial) [inline code/game/machinery/dna_scanner.dm]: 🤨

**Review by Archanial — CHANGES_REQUESTED** [review]

File: code/game/machinery/dna_scanner.dm (hunk 6 above)
Comment (chloecat0) [reply code/game/machinery/dna_scanner.dm]: This just means that there's a 20% chance of getting a mutation and 90% of the time it's negative

File: code/game/machinery/computer/dna_console.dm (hunk 4 above)
Comment (chloecat0) [reply code/game/machinery/computer/dna_console.dm:2061]: Every other emag_act uses |= here

**Pull request merged**